File: cache_interceptor/src/main/java/uk/co/glass_software/android/cache_interceptor/interceptors/cache/SerialisationManager.java
package uk.co.glass_software.android.cache_interceptor.interceptors.cache;

import android.support.annotation.Nullable;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;

import org.iq80.snappy.Snappy;

import uk.co.glass_software.android.cache_interceptor.response.ResponseMetadata;
import uk.co.glass_software.android.cache_interceptor.utils.Action;
import uk.co.glass_software.android.cache_interceptor.utils.Function;
import uk.co.glass_software.android.cache_interceptor.utils.Logger;
import uk.co.glass_software.android.shared_preferences.StoreEntryFactory;

class SerialisationManager {
    
    private final Logger logger;
    private final StoreEntryFactory storeEntryFactory;
    private final boolean encryptData;
    private final boolean compressData;
    private final Gson gson;
    
    SerialisationManager(Logger logger,
                         StoreEntryFactory storeEntryFactory,
                         boolean encryptData,
                         boolean compressData,
                         Gson gson) {
        this.logger = logger;
        this.storeEntryFactory = storeEntryFactory;
        this.encryptData = encryptData;
        this.compressData = compressData;
        this.gson = gson;
    }
    
    @Nullable
    <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> R deserialise(Class<R> responseClass,
                                                                                                        byte[] data,
                                                                                                        Action onError) {
        String simpleName = responseClass.getSimpleName();
        
        try {
            byte[] uncompressed;
            
            if (compressData) {
                uncompressed = Snappy.uncompress(data, 0, data.length);
                logCompression(data, simpleName, uncompressed);
            }
            else {
                uncompressed = data;
            }
            
            if (encryptData) {
                uncompressed = storeEntryFactory.decrypt(uncompressed);
            }
            
            return gson.fromJson(new String(uncompressed), responseClass);
        }
        catch (JsonSyntaxException e) {
            logger.e(this,
                     e,
                     "Cached data didn't match "
                     + simpleName
                     + ": flushing cache"
            );
            onError.act();
            return null;
        }
        catch (Exception e) {
            logger.e(this,
                     e,
                     "Could not deserialise "
                     + simpleName
                     + ": flushing cache"
            );
            onError.act();
            return null;
        }
    }
    
    <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> byte[] serialise(R response) {
        String simpleName = response.getClass().getSimpleName();
        
        byte[] data = gson.toJson(response).getBytes();
        
        if (encryptData) {
            data = storeEntryFactory.encrypt(data);
        }
        
        if (compressData) {
            byte[] compressed = Snappy.compress(data);
            logCompression(compressed, simpleName, data);
            data = compressed;
        }
        
        return data;
    }
    
    private void logCompression(byte[] compressedData,
                                String simpleName,
                                byte[] uncompressed) {
        logger.d(this,
                 "Compressed/uncompressed "
                 + simpleName
                 + ": "
                 + compressedData.length
                 + "B/"
                 + uncompressed.length
                 + "B ("
                 + (100 * compressedData.length / uncompressed.length)
                 + "%)"
        );
    }
}

File: cache_interceptor/src/main/java/uk/co/glass_software/android/cache_interceptor/interceptors/cache/SerialisationManager.kt
package uk.co.glass_software.android.cache_interceptor.interceptors.cache

import com.google.gson.Gson
import org.iq80.snappy.Snappy
import uk.co.glass_software.android.boilerplate.log.Logger
import uk.co.glass_software.android.cache_interceptor.response.ResponseWrapper
import uk.co.glass_software.android.shared_preferences.StoreEntryFactory

internal class SerialisationManager<E>(private val logger: Logger,
                                       private val storeEntryFactory: StoreEntryFactory,
                                       private val gson: Gson)
        where E : Exception,
              E : (E) -> kotlin.Boolean {

    fun deserialise(responseClass: Class<*>,
                    data: ByteArray,
                    isEncrypted: Boolean,
                    isCompressed: Boolean,
                    onError: () -> Unit): ResponseWrapper<E>? {
        val simpleName = responseClass.simpleName

        try {
            var uncompressed = if (isCompressed)
                Snappy.uncompress(data, 0, data.size).apply {
                    logCompression(data, simpleName, this)
                }
            else data

            val encryptionManager = storeEntryFactory.encryptionManager
            if (isEncrypted && encryptionManager != null) {
                uncompressed = encryptionManager.decryptBytes(uncompressed, DATA_TAG)
            }

            val response = gson.fromJson(String(uncompressed), responseClass)
            return ResponseWrapper(responseClass, response, null)
        } catch (e: Exception) {
            logger.e(e, "Could not deserialise $simpleName: clearing the cache")
            onError()
            return null
        }
    }

    fun serialise(response: Any,
                  encryptData: Boolean,
                  compressData: Boolean): ByteArray? {
        val simpleName = response::class.java.simpleName
        val encryptionManager = storeEntryFactory.encryptionManager

        return gson.toJson(response)
                .toByteArray()
                .let {
                    if (encryptData && encryptionManager != null) encryptionManager.encryptBytes(it, DATA_TAG)
                    else it
                }
                ?.let {
                    if (compressData) Snappy.compress(it).also { compressed ->
                        logCompression(compressed, simpleName, it)
                    }
                    else it
                }
    }

    private fun logCompression(compressedData: ByteArray,
                               simpleName: String,
                               uncompressed: ByteArray) {
        logger.d("Compressed/uncompressed $simpleName: ${compressedData.size}B/${uncompressed.size}B (${100 * compressedData.size / uncompressed.size}%)")
    }

    companion object {
        private const val DATA_TAG = "DATA_TAG"
    }
}
